sessiontxn/isolation/repeatable_read.go

```diff
@@ -0,0 +1,267 @@
+	if err := p.baseTxnContextProvider.OnStmtStart(ctx); err != nil {
+		return err
+	}
+
+	p.forUpdateTS = 0
+	p.followingStmtIsPointGetForUpdate = false
+
+	return nil
+}
+
+// OnStmtRetry is the hook that should be called when a statement is retried internally.
+func (p *PessimisticRRTxnContextProvider) OnStmtRetry(ctx context.Context) (err error) {
+	if err = p.baseTxnContextProvider.OnStmtRetry(ctx); err != nil {
+		return err
+	}
+
+	txnCtxForUpdateTS := p.sctx.GetSessionVars().TxnCtx.GetForUpdateTS()
+	// If TxnCtx.forUpdateTS is updated in OnStmtErrorForNextAction, we assign the value to the provider
+	if txnCtxForUpdateTS > p.forUpdateTS {
+		p.forUpdateTS = txnCtxForUpdateTS
+	} else {
+		p.forUpdateTS = 0
+	}
+
+	p.followingStmtIsPointGetForUpdate = false
+
+	return nil
+}
+
+// OnStmtErrorForNextAction is the hook that should be called when a new statement get an error
+func (p *PessimisticRRTxnContextProvider) OnStmtErrorForNextAction(point sessiontxn.StmtErrorHandlePoint, err error) (sessiontxn.StmtErrorAction, error) {
+	switch point {
+	case sessiontxn.StmtErrAfterPessimisticLock:
+		return p.handleAfterPessimisticLockError(err)
+	default:
+		return sessiontxn.NoIdea()
+	}
+}
+
+// Advise is used to give advice to provider
+func (p *PessimisticRRTxnContextProvider) Advise(tp sessiontxn.AdviceType, val []any) error {
+	switch tp {
+	case sessiontxn.AdviceWarmUp:
+		return p.warmUp()
+	case sessiontxn.AdviceOptimizeWithPlan:
+		return p.optimizeWithPlan(val)
+	default:
+		return p.baseTxnContextProvider.Advise(tp, val)
+	}
+}
+
+// optimizeWithPlan optimizes for-update-like point get execution. We set p.followingStmtIsPointGetForUpdate
+// to be true in these cases. And when p.getForUpdateTs is called, we do not acquire tso from PD immediately but return the value of
+// txnCtx.GetForUpdateTS() expecting that the values that the point get operation acquires have not been changed.
+// If these values that the point get operations acquires have been changed, we update ts in error handling.
+func (p *PessimisticRRTxnContextProvider) optimizeWithPlan(val []any) (err error) {
+	if p.isTidbSnapshotEnabled() || p.isBeginStmtWithStaleRead() {
+		return nil
+	}
+
+	if len(val) == 0 {
+		return nil
+	}
+
+	plan, ok := val[0].(plannercore.Plan)
+	if !ok {
+		return nil
+	}
+
+	if execute, ok := plan.(*plannercore.Execute); ok {
+		plan = execute.Plan
+	}
+
+	mayOptimizeForPointGet := false
+	if v, ok := plan.(*plannercore.PhysicalLock); ok {
+		if _, ok := v.Children()[0].(*plannercore.PointGetPlan); ok {
+			mayOptimizeForPointGet = true
+		}
+	} else if v, ok := plan.(*plannercore.Update); ok {
+		if _, ok := v.SelectPlan.(*plannercore.PointGetPlan); ok {
+			mayOptimizeForPointGet = true
+		}
+	} else if v, ok := plan.(*plannercore.Delete); ok {
+		if _, ok := v.SelectPlan.(*plannercore.PointGetPlan); ok {
+			mayOptimizeForPointGet = true
+		}
+	}
+
+	if p.forUpdateTS == 0 {
```

Comment:
I think we can move this if to the begging of the method. When forUpdateTS == 0, do nothing

Reply:
Now, I think we do not this if here.